components/brave_new_tab_ui/reducers/rewards_reducer.ts

```diff
@@ -0,0 +1,189 @@
+
+// API
+import { InitialRewardsData, PreInitialRewardsData } from '../api/initialData'
+import * as storage from '../storage'
+import { getTotalContributions } from '../rewards-utils'
+
+const initialState = storage.load()
+
+export const rewardsReducer: Reducer<NewTab.State | undefined> = (state: NewTab.State | undefined, action: any) => {
+  if (state === undefined) {
+    state = initialState
+  }
+
+  const startingState = state
+  const payload = action.payload
+  switch (action.type) {
+    case types.CREATE_WALLET:
+      chrome.braveRewards.createWallet()
+      state = { ...state }
+      state.rewardsState.walletCreating = true
+      break
+
+    case types.ON_ENABLED_MAIN:
+      state = { ...state }
+      state.rewardsState.enabledMain = payload.enabledMain
+      break
+
+    case types.ON_WALLET_INITIALIZED: {
+      const result: NewTab.RewardsResult = payload.result
+      state = { ...state }
+
+      switch (result) {
+        case NewTab.RewardsResult.WALLET_CORRUPT:
+          state.rewardsState.walletCorrupted = true
+          break
+        case NewTab.RewardsResult.WALLET_CREATED:
+          state.rewardsState.walletCreated = true
+          state.rewardsState.walletCreateFailed = false
+          state.rewardsState.walletCreating = false
+          state.rewardsState.walletCorrupted = false
+          chrome.braveRewards.saveAdsSetting('adsEnabled', 'true')
+          break
+        case NewTab.RewardsResult.LEDGER_OK:
+          state.rewardsState.walletCreateFailed = true
+          state.rewardsState.walletCreating = false
+          state.rewardsState.walletCreated = false
+          state.rewardsState.walletCorrupted = false
+          break
+      }
+      break
+    }
+
+    case types.ON_REWARDS_SETTING_SAVE:
+      state = { ...state }
+
+      const key = action.payload.key
+      const value = action.payload.value
+
+      if (key) {
+        state.rewardsState[key] = !!value
+        chrome.braveRewards.saveSetting(key, value)
+      }
+      break
+
+    case types.ON_ADS_ENABLED:
+      state = { ...state }
+      state.rewardsState.enabledAds = payload.enabled
+      break
+
+    case types.ON_ADS_ESTIMATED_EARNINGS:
+      state = { ...state }
+      state.rewardsState.adsEstimatedEarnings = payload.amount
+      break
+
+    case types.ON_BALANCE_REPORTS:
+      state = { ...state }
+      const reports = payload.reports || {}
+      state.rewardsState.totalContribution = getTotalContributions(reports)
+      break
+
+    case types.DISMISS_NOTIFICATION:
+      state = { ...state }
+
+      const dismissedNotifications = state.rewardsState.dismissedNotifications
+      dismissedNotifications.push(payload.id)
+      state.rewardsState.dismissedNotifications = dismissedNotifications
+
+      state.rewardsState.grants = state.rewardsState.grants.filter((grant) => {
+        return grant.promotionId !== payload.id
+      })
+      break
+
+    case types.ON_GRANT:
+      state = { ...state }
```

Comment:
same comment re: creating a new state when not changing any properties

Reply:
done